Add RBAC for Windows desktop access

zmb3 opened the pull request:
This commit adds RBAC checks for Windows Desktops as described in
RFD 33 and RFD 34:

- add Windows desktop logins & labels to role definition
- introduce new file config for host labels based on a regexp match
- auth server API performs access checking for Windows desktop resources
- add RDP client callback to authorize the user
- support user/role locks
- respect the client idle timeout setting

Note: in cases where an connection is terminated to to RBAC, the web UI
currently displays "websocket connection failed" because the connection
is closed from the server. We'll need to follow up with a nice error
message for the client side to improve the UX here.

Updates #7761

zmb3 committed:
Add RBAC for Windows desktop access

This commit adds RBAC checks for Windows Desktops as described in
RFD 33 and RFD 34:

- add Windows desktop logins & labels to role definition
- introduce new file config for host labels based on a regexp match
- auth server API performs access checking for Windows desktop resources
- add RDP client callback to authorize the user
- support user/role locks
- respect the client idle timeout setting

Note: in cases where an connection is terminated to to RBAC, the web UI
currently displays "websocket connection failed" because the connection
is closed from the server. We'll need to follow up with a nice error
message for the client side to improve the UX here.

Updates #7761

zmb3 commented:
Leaving this in draft state while I resolve merge conflicts, but if I've added you as a reviewer and you want to take a look feel free to start reviewing.

zmb3 committed:
Merge branch 'master' into zmb3/windows-rbac

zmb3 committed:
Merge branch 'master' into zmb3/windows-rbac

awly reviewed (commented):
Are the `webassets` changes here intentional?

codingllama reviewed (commented):
There are a bunch of concurrent reviews happening, so sorry for eventual duplicate comments.

I'm missing test for most `auth_with_roles.go` changes, could we get some coverage there?

zmb3 committed:
Address review feedback

zmb3 committed:
Merge branch 'master' into zmb3/windows-rbac

zmb3 committed:
fix webassets

zmb3 committed:
Ensure Admin and WindowsDesktop roles have access to all labels

zmb3 committed:
Remove OSS RBAC migration marked for deletion

This included limits to the default admin role, which was preventing
the wildcard Windows desktop labesl from showing up.

zmb3 committed:
Merge branch 'master' into zmb3/windows-rbac

zmb3 committed:
Stop creating a default admin role

With this change, we no longer create a default admin role as it
is too powerful. Instead we prefer the "preset" edit, audit, and
access roles.

Also add wildcard labels for Windows desktop access to the preset
access role. Note: this only applies for new clusters - teleport
will not overwrite the access role on existing clusters to avoid
clobbering user modifications.

zmb3 committed:
Use teleport.dev/ prefix for automatically applied labels

zmb3 committed:
Merge branch 'master' into zmb3/windows-rbac

zmb3 committed:
Remove unused admin role setup code

zmb3 committed:
Address review feedback

zmb3 committed:
Add RBAC to desktop access docs

zmb3 committed:
Be more explicit with regexp matching the beginning of the string

zmb3 committed:
Link release build for Rust lib on darwin, too

zmb3 committed:
Merge branch 'master' into zmb3/windows-rbac

zmb3 committed:
Fix a few tests that assume an admin role is always present

zmb3 committed:
Merge branch 'master' into zmb3/windows-rbac
